pkg/stanza/fileconsumer/file.go

```diff
@@ -266,19 +259,11 @@ func (m *Manager) clearCurrentFingerprints() {
 // known files, then increments the generation of all tracked old readers
 // before clearing out readers that have existed for 3 generations.
 func (m *Manager) saveCurrent(readers []*reader) {
-	// Add readers from the current, completed poll interval to the list of known files
-	m.knownFiles = append(m.knownFiles, readers...)
-
-	// Clear out old readers. They are sorted such that they are oldest first,
-	// so we can just find the first reader whose generation is less than our
-	// max, and keep every reader after that
-	for i := 0; i < len(m.knownFiles); i++ {
-		reader := m.knownFiles[i]
-		if reader.generation <= 3 {
-			m.knownFiles = m.knownFiles[i:]
-			break
-		}
+	forgetNum := len(m.knownFiles) + len(readers) - cap(m.knownFiles)
+	if forgetNum > 0 {
+		m.knownFiles = append(m.knownFiles[forgetNum:], readers...)
 	}
+	m.knownFiles = append(m.knownFiles, readers...)
 }
```

Comment:
Should this be like the following? Because we append same set of readers twice if `forgetNum` exceeds `0`
```suggestion
	forgetNum := len(m.knownFiles) + len(readers) - cap(m.knownFiles)
	if forgetNum > 0 {
		m.knownFiles = append(m.knownFiles[forgetNum:], readers...)
	} else {
		m.knownFiles = append(m.knownFiles, readers...)
	}
}
```

Reply:
Yes, good catch

Reply:
@djaglowski after your latest commit, I still see that `else` is missing.

Reply:
Yes, I added a return from the `if`

Reply:
huh missed that, my bad. changes look good otherwise!